backend/controller/cronjobs/cronjobs_test.go

```diff
@@ -0,0 +1,225 @@
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	job := dal.CronJob{
+		Key:           model.NewCronJobKey(module, verb),
+		DeploymentKey: deploymentKey,
+		Ref:           schema.Ref{Module: module, Name: verb},
+		Schedule:      schedule,
+		StartTime:     startTime,
+		NextExecution: nextExecution,
+		State:         dal.JobStateIdle,
+	}
+	d.jobs = append(d.jobs, job)
+}
+
+func (d *mockDAL) indexForJob(job dal.CronJob) (int, error) {
+	for i, j := range d.jobs {
+		if j.Key.String() == job.Key.String() {
+			return i, nil
+		}
+	}
+	return -1, fmt.Errorf("job not found")
+}
+
+func (d *mockDAL) StartCronJobs(ctx context.Context, jobs []dal.CronJob) (attemptedJobs []dal.AttemptedCronJob, err error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	attemptedJobs = []dal.AttemptedCronJob{}
+	now := (*d.clock).Now()
+
+	for _, inputJob := range jobs {
+		i, err := d.indexForJob(inputJob)
+		if err != nil {
+			return nil, err
+		}
+		job := d.jobs[i]
+		if !job.NextExecution.After(now) && job.State == dal.JobStateIdle {
+			job.State = dal.JobStateExecuting
+			job.StartTime = (*d.clock).Now()
+			d.jobs[i] = job
+			attemptedJobs = append(attemptedJobs, dal.AttemptedCronJob{
+				CronJob:           job,
+				DidStartExecution: true,
+				HasMinReplicas:    true,
+			})
+		} else {
+			attemptedJobs = append(attemptedJobs, dal.AttemptedCronJob{
+				CronJob:           job,
+				DidStartExecution: false,
+				HasMinReplicas:    true,
+			})
+		}
+		d.attemptCountMap[job.Key.String()]++
+	}
+	return attemptedJobs, nil
+}
+
+func (d *mockDAL) EndCronJob(ctx context.Context, job dal.CronJob, next time.Time) (dal.CronJob, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	i, err := d.indexForJob(job)
+	if err != nil {
+		return dal.CronJob{}, err
+	}
+	internalJob := d.jobs[i]
+	if internalJob.State != dal.JobStateExecuting {
+		return dal.CronJob{}, fmt.Errorf("job can not be stopped, it isnt running")
+	}
+	if internalJob.StartTime != job.StartTime {
+		return dal.CronJob{}, fmt.Errorf("job can not be stopped, start time does not match")
+	}
+	internalJob.State = dal.JobStateIdle
+	internalJob.NextExecution = next
+	d.jobs[i] = internalJob
+	return internalJob, nil
+}
+
+func (d *mockDAL) GetStaleCronJobs(ctx context.Context, duration time.Duration) ([]dal.CronJob, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return slices.Filter(d.jobs, func(job dal.CronJob) bool {
+		return (*d.clock).Now().After(job.StartTime.Add(duration))
+	}), nil
+}
+
+type mockScheduler struct {
+}
+
+func (s *mockScheduler) Singleton(retry backoff.Backoff, job scheduledtask.Job) {
+	// do nothing
+}
+
+func (s *mockScheduler) Parallel(retry backoff.Backoff, job scheduledtask.Job) {
+	// do nothing
+}
+
+type controller struct {
+	key      model.ControllerKey
+	DAL      DAL
+	cronJobs *Service
+}
+
+func TestService(t *testing.T) {
```

Comment:
Does this test the hashring? It's a bit hard for me to tell.

Reply:
It did, but not well. I've added another unit test focused on that.